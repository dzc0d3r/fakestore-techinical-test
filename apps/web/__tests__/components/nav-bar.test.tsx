// __tests__/components/nav-bar.test.tsx
import NaVBar from "@/components/navbar";
import { act, render } from "@testing-library/react";
import React from "react";
import { describe, expect, it, vi } from "vitest";

// Mock authentication functions
vi.mock("@/auth", () => ({
  auth: vi.fn().mockResolvedValue({ user: { role: "user" } }),
  signOut: vi.fn(),
}));

// Mock Next.js Image to prevent issues with Next/Image
vi.mock("next/image", () => ({
  __esModule: true,
  default: (props: any) => <img {...props} alt={props.alt || "mocked image"} />,
}));

// Mock the Cart component to avoid internal hook errors
vi.mock("@/components/navbar/cart", () => ({
  default: () => <div data-testid="cart">Cart Component</div>,
}));

// Mock shadcn / Radix UI based components with simple wrappers
vi.mock("@/components/ui/button", () => ({
  Button: (props: any) => <button {...props}>{props.children}</button>,
}));

vi.mock("@/components/ui/dropdown-menu", () => ({
  DropdownMenu: ({ children }: { children: React.ReactNode }) => (
    <div>{children}</div>
  ),
  DropdownMenuContent: ({ children }: { children: React.ReactNode }) => (
    <div>{children}</div>
  ),
  DropdownMenuItem: ({ children }: { children: React.ReactNode }) => (
    <div>{children}</div>
  ),
  DropdownMenuTrigger: ({ children }: { children: React.ReactNode }) => (
    <div>{children}</div>
  ),
  DropdownMenuSeparator: () => <div />,
}));

vi.mock("@/components/ui/sheet", () => ({
  Sheet: ({ children }: { children: React.ReactNode }) => <div>{children}</div>,
  SheetContent: ({ children }: { children: React.ReactNode }) => (
    <div>{children}</div>
  ),
  SheetTrigger: ({ children }: { children: React.ReactNode }) => (
    <div>{children}</div>
  ),
}));

vi.mock("@/components/ui/input", () => ({
  Input: (props: any) => <input {...props} />,
}));

// Mock LoginButton if needed
vi.mock("@/components/navbar/login-button", () => ({
  default: () => <div data-testid="login-button">Login Button</div>,
}));

describe("NaVBar Snapshot", () => {
  it("should render correctly and match snapshot", async () => {
    let element: JSX.Element | null = null;

    // Resolve the async component using act
    await act(async () => {
      element = await NaVBar();
    });

    const { container } = render(element!);

    // Remove the invalid "action" attribute from any <form> elements to avoid warnings
    container.querySelectorAll("form[action]").forEach((form) => {
      form.removeAttribute("action");
    });

    expect(container).toMatchSnapshot();
  });
});
