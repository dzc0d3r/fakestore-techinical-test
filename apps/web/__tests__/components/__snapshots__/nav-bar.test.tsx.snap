// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NaVBar Snapshot > should render correctly and match snapshot 1`] = `
<div>
  <header
    class="sticky backdrop-blur-lg bg-white backdrop-filter  bg-opacity-80 top-0 shadow-sm flex h-16 items-center gap-4 px-4 md:px-6 z-50"
  >
    <nav
      class="hidden flex-col gap-2 text-lg font-medium md:flex md:flex-row md:items-center md:gap-5 md:text-sm lg:gap-2"
    >
      <a
        class="flex items-center gap-2 text-lg font-semibold md:text-base"
        href="/"
      >
        <img
          alt="FakeStore Inc"
          height="120"
          placeholder="blur"
          src="/src/public/logo.png"
          width="120"
        />
        <span
          class="sr-only"
        >
          Acme Inc
        </span>
      </a>
      <a
        class="text-primary/80 hover:text-foreground"
        href="/about"
      >
        <button
          variant="ghost"
        >
          About
        </button>
      </a>
      <a
        class="text-primary/80 hover:text-foreground"
        href="privacy"
      >
        <button
          variant="ghost"
        >
          Privacy
        </button>
      </a>
    </nav>
    <div>
      <div>
        <button
          class="shrink-0 md:hidden"
          variant="ghost"
        >
          <svg
            class="lucide lucide-align-left h-5 w-5"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15 12H3"
            />
            <path
              d="M17 18H3"
            />
            <path
              d="M21 6H3"
            />
          </svg>
          <span
            class="sr-only"
          >
            Toggle navigation menu
          </span>
        </button>
      </div>
      <div>
        <nav
          class="grid gap-6 text-lg font-medium"
        >
          <a
            class="flex items-center gap-2 text-lg font-semibold"
            href="/"
          >
            <img
              alt="FakeStore Inc"
              height="50"
              placeholder="blur"
              src="/src/public/logo.png"
              width="50"
            />
            <span
              class="sr-only"
            >
              Acme Inc
            </span>
          </a>
          <div
            class="grid gap-4 mt-5"
          >
            <a
              class="text-primary/80 hover:text-foreground"
              href="/about"
            >
              About
            </a>
            <a
              class="text-primary/80 hover:text-foreground"
              href="privacy"
            >
              Privacy
            </a>
          </div>
        </nav>
      </div>
    </div>
    <div
      class="flex w-full items-center gap-4 md:ml-auto md:gap-2 lg:gap-4"
    >
      <form
        class="ml-auto flex-1 sm:flex-initial"
      >
        <div
          class="relative"
        >
          <svg
            class="lucide lucide-search absolute left-2.5 top-2.5 h-4 w-4 text-muted-foreground"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="11"
              cy="11"
              r="8"
            />
            <path
              d="m21 21-4.3-4.3"
            />
          </svg>
          <input
            class="pl-8 sm:w-[300px] md:w-[200px] lg:w-[300px]"
            placeholder="Search products..."
            type="search"
          />
        </div>
      </form>
      <div
        data-testid="cart"
      >
        Cart Component
      </div>
      <div>
        <div>
          <button
            class="rounded-full"
            variant="secondary"
          >
            <svg
              class="lucide lucide-circle-user h-5 w-5"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="10"
              />
              <circle
                cx="12"
                cy="10"
                r="3"
              />
              <path
                d="M7 20.662V19a2 2 0 0 1 2-2h6a2 2 0 0 1 2 2v1.662"
              />
            </svg>
            <span
              class="sr-only"
            >
              Toggle user menu
            </span>
          </button>
        </div>
        <div>
          <div>
            <a
              class="flex flex-row gap-2 lg:text-md w-full font-medium"
              href="/profile"
            >
              <svg
                class="lucide lucide-settings  left-2"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12.22 2h-.44a2 2 0 0 0-2 2v.18a2 2 0 0 1-1 1.73l-.43.25a2 2 0 0 1-2 0l-.15-.08a2 2 0 0 0-2.73.73l-.22.38a2 2 0 0 0 .73 2.73l.15.1a2 2 0 0 1 1 1.72v.51a2 2 0 0 1-1 1.74l-.15.09a2 2 0 0 0-.73 2.73l.22.38a2 2 0 0 0 2.73.73l.15-.08a2 2 0 0 1 2 0l.43.25a2 2 0 0 1 1 1.73V20a2 2 0 0 0 2 2h.44a2 2 0 0 0 2-2v-.18a2 2 0 0 1 1-1.73l.43-.25a2 2 0 0 1 2 0l.15.08a2 2 0 0 0 2.73-.73l.22-.39a2 2 0 0 0-.73-2.73l-.15-.08a2 2 0 0 1-1-1.74v-.5a2 2 0 0 1 1-1.74l.15-.09a2 2 0 0 0 .73-2.73l-.22-.38a2 2 0 0 0-2.73-.73l-.15.08a2 2 0 0 1-2 0l-.43-.25a2 2 0 0 1-1-1.73V4a2 2 0 0 0-2-2z"
                />
                <circle
                  cx="12"
                  cy="12"
                  r="3"
                />
              </svg>
               Settings
            </a>
          </div>
          <div />
          <form>
            <button
              class="w-full relative"
              type="submit"
              variant="destructive"
            >
              <svg
                class="lucide lucide-log-out absolute left-3"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M9 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h4"
                />
                <polyline
                  points="16 17 21 12 16 7"
                />
                <line
                  x1="21"
                  x2="9"
                  y1="12"
                  y2="12"
                />
              </svg>
               Logout
            </button>
          </form>
        </div>
      </div>
    </div>
  </header>
</div>
`;
